lib/es6/activities/baseIterator.js

```diff
@@ -0,0 +1,130 @@
+const _ = require('lodash');
+const { errors } = require('../common');
+const Activity = require('./activity')
+const Block = require('./block')
+const WithBody = require('./withBody')
+const { LoopBreakEvent, LoopContinueEvent } = require('../common/events');
+
+class BaseIterator extends Activity {
+  constructor() {
+    super();
+    this.items = null;
+    this.varName = 'item';
+    this.parallel = false;
+    this._bodies = null;
+  }
+  initializeStructure() {
+    if (this.parallel) {
+      const numCPUs = require('os').cpus().length;
+      this._bodies = [];
+      if (this.args && this.args.length) {
+        /* eslint-disable no-plusplus */
+        for (let i = 0; i < Math.min(process.env.UV_THREADPOOL_SIZE || 100000, numCPUs); i++) {
+          const newArgs = [];
+          for (const arg of this.args) {
+            if (arg instanceof Activity) {
+              newArgs.push(arg.clone());
+            } else {
+              newArgs.push(arg);
+            }
+          }
+          const newBody = new Block();
+          newBody.args = newArgs;
+          this._bodies.push(newBody);
+        }
+      }
+      this.args = null;
+    } else {
+      WithBody.prototype.initializeStructure.call(this);
+    }
+  }
+
+  run(callContext) {
+    const varName = this.varName;
+    const items = this.items;
+    if (!_.isNull(items)) {
+      this[varName] = null;
+      callContext.schedule(items, '_itemsGot');
+    } else {
+      callContext.complete();
+    }
+  }
+
+  _itemsGot(callContext, reason, result) {
+    if (reason === Activity.states.complete && !_.isUndefined(result)) {
+      if (result && _.isFunction(result.next)) {
+        this._iterator = result;
+      } else {
+        this._remainingItems = _.isArray(result) ? result.slice() : [result];
+      }
+      callContext.activity._doStep.call(this, callContext);
+    } else {
+      callContext.end(reason, result);
+    }
+  }
+
+  _doStep(callContext, lastResult) {
+    const varName = this.varName;
+    const remainingItems = this._remainingItems;
+    const iterator = this._iterator;
+    if (remainingItems && remainingItems.length) {
+      if (this.parallel) {
+        const bodies = this._bodies;
+        const pack = [];
+        let idx = 0;
+        while (remainingItems.length && idx < bodies.length) {
+          const item = remainingItems[0];
+          remainingItems.splice(0, 1);
```

Comment:
You could also use the splice() return value and remove the previous statement. Another option is to use remainingItems.shift() to remove the head element.

Reply:
I agree, but I'm going to leave as-is - reason being that this was a refactor/port of W4N `@For` code, and I don't really want to mess with it.